core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java

```diff
@@ -80,6 +81,16 @@ final class DefaultClientFactory implements ClientFactory {
             try {
                 Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                     if (!shutdownHookDisabled) {
+                        final long gracefulShutdownTimeoutMillis =
+                                Flags.defaultClientFactoryGracefulShutdownTimeoutMillis();
+                        if (gracefulShutdownTimeoutMillis > 0) {
+                            try {
+                                Thread.sleep(gracefulShutdownTimeoutMillis);
```

Comment:
- Shouldn't we allow a user specify the graceful shutdown timeout for each `ClientFactory`? This PR adds the sleep only to the default factory, which doesn't sound right.
- What should we do when a user sends a request during this period? Should we reject them?
- Can we skip waiting when there are no pending requests? Do we need grace period just like we have on the server side shutdown? Just adding a fixed amount of sleep doesn't sound right.

Reply:
> Shouldn't we allow a user specify the graceful shutdown timeout for each ClientFactory?

Good idea.

> Can we skip waiting when there are no pending requests? Do we need grace period just like we have on the server side shutdown? Just adding a fixed amount of sleep doesn't sound right.
If the server allows new requests for 10 seconds considering inflight requests and health check interval, the client should not be closed even if there are no active requests on the client.

> What should we do when a user sends a request during this period? Should we reject them?

Since the server controls incoming traffic through health checks, I am not sure if it is right for the client to reject new requests.

The more I think about it, the problem cannot be solved without linking the lifecycles of the server and client.